verification/utils/slices/slices_contents.gobra

```diff
@@ -0,0 +1,194 @@
+	// start inclusive
+	0 <= start &&
+	start <= end &&
+	// end exclusive
+	end <= cap(s) &&
+	len(contents) == end - start &&
+	// Maybe, it is worth having different versions with
+	// different triggers, or using a different trigger
+	// than the one that is inferred.
+	forall i int :: start <= i && i < end ==> (acc(&s[i]) && s[i] == contents[i - start])
+}
+
+ghost
+requires 0 < p
+requires 0 <= start && start <= end && end <= len(s)
+requires acc(AbsSlice_Bytes(s, start, end), p)
+ensures  acc(AbsSliceC_Bytes(s, start, end, contents), p)
+decreases
+func GetAbsSliceC_Bytes(s []byte, start int, end int, p perm) (contents seq[byte]) {
+	contents = NewSeq_Bytes(end - start)
+	unfold acc(AbsSlice_Bytes(s, start, end), p)
+	invariant forall k int :: { &s[k] } start <= k && k < end ==> acc(&s[k], p)
+	invariant start <= i && i <= end
+	invariant len(contents) == end - start
+	invariant forall k int :: { contents[k] } start <= k && k < i ==> contents[k - start] == s[k]
+	decreases end - i
+	for i := start; i < end; i += 1 {
+		contents[i - start] = s[i]
+	}
+	fold acc(AbsSliceC_Bytes(s, start, end, contents), p)
+	return contents
+}
+
+ghost
+requires 0 <= p
+requires AbsSliceC_Bytes(s, start, end, contents)
+requires start < idx && idx < end
+ensures  AbsSliceC_Bytes(s, start, idx, contents[:idx-start]) && AbsSliceC_Bytes(s, idx, end, contents[idx-start:])
+decreases
+func SplitByIndexC_Bytes(s []byte, start int, end int, idx int, contents seq[byte], p perm) {
+	unfold AbsSliceC_Bytes(s, start, end, contents)
+	assert len(contents) == end - start
+	fold   AbsSliceC_Bytes(s, start, idx, contents[:idx-start])
+	fold   AbsSliceC_Bytes(s, idx, end, contents[idx-start:])
+}
+
+ghost
+requires 0 < p
+requires acc(AbsSliceC_Bytes(s, start, idx, contents1), p)
+requires acc(AbsSliceC_Bytes(s, idx,   end, contents2), p)
+ensures  acc(AbsSliceC_Bytes(s, start, end, contents1 ++ contents2), p)
+decreases
+func CombineAtIndexC_Bytes(s []byte, start int, end int, idx int, contents1 seq[byte], contents2 seq[byte], p perm) {
+	unfold acc(AbsSliceC_Bytes(s, start, idx, contents1), p)
+	unfold acc(AbsSliceC_Bytes(s, idx,   end, contents2), p)
+	fold   acc(AbsSliceC_Bytes(s, start, end, contents1 ++ contents2), p)
+}
+
+ghost
+requires AbsSliceC_Bytes(s, start, end, contents)
+// the following precondition convinces Gobra that
+// the slice operation is well-formed
+requires unfolding AbsSliceC_Bytes(s, start, end, contents) in true
+ensures  AbsSliceC_Bytes(s[start:end], 0, len(s[start:end]), contents)
+decreases
+func ResliceC_Bytes(s []byte, start int, end int, contents seq[byte]) {
+	unfold AbsSliceC_Bytes(s, start, end, contents)
+	assert forall i int :: 0 <= i && i < (end-start) ==> &s[start:end][i] == &s[start + i]
+	fold  AbsSliceC_Bytes(s[start:end], 0, len(s[start:end]), contents)
+}
+
+
+// The slice axiomatization is very incomplete.
+// The following body cannot be verified yet.
```

Comment:
If this does not verify, let's not make it available (i.e., let's comment also the signature). Did you check whether it would verify with, for example, more & better triggers?

Reply:
we don't need it yet anyway since I changed direction so I will just remove it and we will introduce it if we ever need it